### PR TITLE
Don't repeat lifetime names from outer binder in print

### LINKED ISSUE #102392: Spurious additional lifetime appearing in printout of type involving HRT lifetimes
“Related” to #102346, and #101280, as this, too, is about diagnostic printing of types with HRT lifetimes (unlike the former related issue, this issue is not a recent regression):
```rs
fn g(f: for<'a> fn(fn(&str, &'a str))) -> bool {
    f
}
```
```
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
1 | fn g(f: for<'a> fn(fn(&str, &'a str))) -> bool {
  |                                           ---- expected `bool` because of return type
2 |     f
  |     ^ expected `bool`, found fn pointer
  |
  = note:    expected type `bool`
          found fn pointer `for<'a> fn(for<'b, 'a> fn(&'b str, &'a str))`
```

The way this type is displayed includes an additional superflous/confusing/wrong quantification of `'a` in the inner `for`. Correct would be `for<'a> fn(for<'b> fn(&'b str, &'a str))`, not `for<'a> fn(for<'b, 'a> fn(&'b str, &'a str))`.

cc @b-naber who is currently working on the above-mentioned “related” issues.

@rustbot label regression-from-stable-to-stable, E-needs-bisection

Regresses between 1.54 and 1.55:

1.54
```
  = note:    expected type `bool`
          found fn pointer `for<'a> fn(for<'r> fn(&'r str, &'a str))`
```
1.55
```
  = note:    expected type `bool`
          found fn pointer `for<'a> fn(for<'r, 'a> fn(&'r str, &'a str))`
```

### LINKED ISSUE #102414: Duplicated lifetime name in diagnostic for higher-ranked type in higher-ranked trait bound
In https://github.com/rust-lang/rust/pull/101454/files#diff-46c19c4c60dbf4dc95fd23bdb31dbba31d7001a7f269c4068c536522a09f25deR1

The diagnostic looks a bit silly. I don't know why we are choosing a `'b` for the inner param.

_Originally posted by @cjgillot in https://github.com/rust-lang/rust/pull/101454#discussion_r963015967_

### PR DESCRIPTION
**Pull request by b-naber**
Fixes https://github.com/rust-lang/rust/issues/102392
Fixes https://github.com/rust-lang/rust/issues/102414

r? @lcnr

### CODE REVIEW

File: src/test/ui/where-clauses/higher-ranked-fn-type.quiet.stderr [hunk 1]
```diff
@@ -1,8 +1,8 @@
-error[E0277]: the trait bound `for<'b> for<'b> fn(&'b ()): Foo` is not satisfied
+error[E0277]: the trait bound `for<'b> fn(&'b ()): Foo` is not satisfied
   --> $DIR/higher-ranked-fn-type.rs:20:5
    |
 LL |     called()
-   |     ^^^^^^ the trait `for<'b> Foo` is not implemented for `for<'b> fn(&'b ())`
+   |     ^^^^^^ the trait `for<'b> Foo` is not implemented for `fn(&'b ())`
```
Comment (b-naber) [inline src/test/ui/where-clauses/higher-ranked-fn-type.quiet.stderr:5]: While this does fix https://github.com/rust-lang/rust/issues/102414, it's kind of unfortunate that this omits the binder here. Not sure why that is yet.

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 2]
```diff
@@ -2234,24 +2246,57 @@ impl<'tcx> FmtPrinter<'_, 'tcx> {
             })
         } else {
             let tcx = self.tcx;
-            let mut name = |br: ty::BoundRegion| {
-                start_or_continue(&mut self, "for<", ", ");
-                let kind = match br.kind {
+            let mut name = |db: Option<ty::DebruijnIndex>,
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:2249]: Could you give `db` a more explicit name?

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 3]
```diff
@@ -2055,7 +2055,14 @@ struct RegionFolder<'a, 'tcx> {
     tcx: TyCtxt<'tcx>,
     current_index: ty::DebruijnIndex,
     region_map: BTreeMap<ty::BoundRegion, ty::Region<'tcx>>,
-    name: &'a mut (dyn FnMut(ty::BoundRegion) -> ty::Region<'tcx> + 'a),
+    name: &'a mut (
+                dyn FnMut(
+        Option<ty::DebruijnIndex>,
+        ty::DebruijnIndex,
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:2061]: Could you add a comment stating what the two `DebruijnIndex` are?

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 4]
```diff
@@ -2086,7 +2093,9 @@ impl<'a, 'tcx> ty::TypeFolder<'tcx> for RegionFolder<'a, 'tcx> {
     fn fold_region(&mut self, r: ty::Region<'tcx>) -> ty::Region<'tcx> {
         let name = &mut self.name;
         let region = match *r {
-            ty::ReLateBound(_, br) => *self.region_map.entry(br).or_insert_with(|| name(br)),
+            ty::ReLateBound(db, br) => {
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_middle/src/ty/print/pretty.rs:2096]: (ah, that's whatyou're currently doing inside of `name`. i think it's better to do this here instead)

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 5]
```diff
@@ -2095,7 +2104,10 @@ impl<'a, 'tcx> ty::TypeFolder<'tcx> for RegionFolder<'a, 'tcx> {
                     _ => {
                         // Index doesn't matter, since this is just for naming and these never get bound
                         let br = ty::BoundRegion { var: ty::BoundVar::from_u32(0), kind };
-                        *self.region_map.entry(br).or_insert_with(|| name(br))
+                        *self
+                            .region_map
+                            .entry(br)
+                            .or_insert_with(|| name(None, self.current_index, br))
```
[the remark this answers is not in the document]
Comment (b-naber) [reply compiler/rustc_middle/src/ty/print/pretty.rs:2110]: Seems like the reason this isn't printed is that we [skip the binder](https://github.com/rust-lang/rust/blob/4bd30785eb6e55f317b41b7c278f438807eeb174/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs#L465) in the error reporting code?!

**Comment by jackh726** [comment]
Its actually crazy because I half-fixed this completely independently locally, when working on something else.

I don't think we should stop printing placeholders as bound vars. They're sort of an "internal detail" of how we solve higher-ranked bounds. What instead would you want to show the user? The lifetime name must come from somewhere.

File: src/test/ui/lifetimes/re-empty-in-error.stderr [hunk 6]
```diff
@@ -4,7 +4,7 @@ error: higher-ranked lifetime error
 LL |     foo(&10);
    |     ^^^^^^^^
    |
-   = note: could not prove `for<'b, 'a> &'b (): 'a`
+   = note: could not prove `for<'b> &'b (): '_`
```
Comment (b-naber) [inline src/test/ui/lifetimes/re-empty-in-error.stderr:7]: With the proposed change we don't create names for all anonymous late-bound vars.

Comment (jackh726) [reply src/test/ui/lifetimes/re-empty-in-error.stderr:7]: Yeah, this is what I did locally. Kind of weird. I guess it makes more sense to do what you had before.

File: compiler/rustc_middle/src/ty/print/pretty.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_middle/src/ty/print/pretty.rs:2110]: yeah, we shouldn't do that. Either `map_bound` or replace with placeholders. Skipping binders is incorrect for errors as well i think

**Comment by jackh726** [comment]
I think this is a good start.

@bors r+

**Pull request merged**